pkg/resources/task.go

```diff
@@ -2,709 +2,648 @@ package resources
-		if err != nil {
-			return fmt.Errorf("error suspending task %s, err: %w", taskId.FullyQualifiedName(), err)
+	if d.HasChange("sql_statement") {
+		if err := client.Tasks.Alter(ctx, sdk.NewAlterTaskRequest(id).WithModifyAs(d.Get("sql_statement").(string))); err != nil {
+			return diag.FromErr(err)
 		}
+	}
 
-		o, n := d.GetChange("after")
-		oldAfter := expandStringList(o.([]interface{}))
-		newAfter := expandStringList(n.([]interface{}))
+	if d.HasChange("finalize") {
+		if v, ok := d.GetOk("finalize"); ok {
+			rootTaskId, err := sdk.ParseSchemaObjectIdentifier(v.(string))
+			if err != nil {
+				return diag.FromErr(err)
+			}
 
-		if len(newAfter) > 0 {
-			// preemptively removing schedule because a task cannot have both after and schedule
-			if err := client.Tasks.Alter(ctx, sdk.NewAlterTaskRequest(taskId).WithUnset(*sdk.NewTaskUnsetRequest().WithSchedule(true))); err != nil {
-				return fmt.Errorf("error updating schedule on task %s", taskId.FullyQualifiedName())
+			rootTask, err := client.Tasks.ShowByID(ctx, rootTaskId)
+			if err != nil {
+				return diag.FromErr(err)
 			}
-		}
 
-		// Remove old dependencies that are not in new dependencies
-		toRemove := make([]sdk.SchemaObjectIdentifier, 0)
-		for _, dep := range oldAfter {
-			if !slices.Contains(newAfter, dep) {
-				toRemove = append(toRemove, sdk.NewSchemaObjectIdentifierInSchema(taskId.SchemaId(), dep))
+			if rootTask.State == sdk.TaskStateStarted {
+				if err := client.Tasks.Alter(ctx, sdk.NewAlterTaskRequest(rootTaskId).WithSuspend(true)); err != nil {
+					return diag.FromErr(sdk.JoinErrors(err))
+				}
 			}
-		}
-		if len(toRemove) > 0 {
-			if err := client.Tasks.Alter(ctx, sdk.NewAlterTaskRequest(taskId).WithRemoveAfter(toRemove)); err != nil {
-				return fmt.Errorf("error removing after dependencies from task %s", taskId.FullyQualifiedName())
+
+			if err := client.Tasks.Alter(ctx, sdk.NewAlterTaskRequest(id).WithSetFinalize(rootTaskId)); err != nil {
+				return diag.FromErr(err)
 			}
-		}
 
-		// Add new dependencies that are not in old dependencies
-		toAdd := make([]sdk.SchemaObjectIdentifier, 0)
-		for _, dep := range newAfter {
-			if !slices.Contains(oldAfter, dep) {
-				toAdd = append(toAdd, sdk.NewSchemaObjectIdentifierInSchema(taskId.SchemaId(), dep))
+			if rootTask.State == sdk.TaskStateStarted && !slices.ContainsFunc(tasksToResume, func(identifier sdk.SchemaObjectIdentifier) bool {
+				return identifier.FullyQualifiedName() == rootTaskId.FullyQualifiedName()
+			}) {
+				tasksToResume = append(tasksToResume, rootTaskId)
 			}
-		}
-		if len(toAdd) > 0 {
-			for _, depId := range toAdd {
-				tasksToResume, err := client.Tasks.SuspendRootTasks(ctx, depId, taskId)
-				defer func() {
-					if err := client.Tasks.ResumeTasks(ctx, tasksToResume); err != nil {
-						log.Printf("[WARN] failed to resume tasks: %s", err)
-					}
-				}()
-				if err != nil {
-					return err
+		} else {
+			if task.TaskRelations.FinalizedRootTask == nil {
```

Comment:
I'm not sure if we should fail here. A user declaratively says in the config that the finalizer must be unset, and I think it's not important that it was unset before.

Reply:
Hmm you're probably right, I'll add a test for this case in this pr and adjust this if needed

Reply:
It seems like test passed, so I guess this is just a safeguard, maybe let's leave this comment for now and I'll take a closer look at it in the next pr. You can take a look at the test I added, imo it looks good, but maybe I did something wrong.

Reply:
Yeah, so this else block is executed when there's a change non-nil->nil. But in this if we expect the current state to be nil, and both of these can't happen at the same time, right?

Reply:
Removed it, it seems from the tests it wasn't needed.